src/rust/engine/src/watch.rs

```diff
@@ -0,0 +1,159 @@
+              .map(|path| {
+                // relativize paths to build root.
+                let mut paths_to_invalidate: Vec<PathBuf> = vec![];
+                let path_relative_to_build_root = {
+                  if path.starts_with(&canonical_build_root) {
+                    path.strip_prefix(&canonical_build_root).unwrap().into()
+                  } else {
+                    path
+                  }
+                };
+                paths_to_invalidate.push(path_relative_to_build_root.clone());
+                if let Some(parent_dir) = path_relative_to_build_root.parent() {
+                  paths_to_invalidate.push(parent_dir.to_path_buf());
+                }
+                paths_to_invalidate
+              })
+              .flatten()
+              .collect();
+            info!("notify invalidating {:?} because of {:?}", paths, ev.kind);
+            InvalidationWatcher::invalidate(&graph, &paths, "notify");
+          }
+          Ok(Err(err)) => {
+            if let notify::ErrorKind::PathNotFound = err.kind {
+              warn!("Path(s) did not exist: {:?}", err.paths);
+              continue;
+            } else {
+              error!("File watcher failing with: {}", err);
+              break;
+            }
+          }
+          Err(RecvTimeoutError::Timeout) => continue,
+          Err(RecvTimeoutError::Disconnected) => {
+            // The Watcher is gone: we're done.
+            break;
+          }
+        };
+      }
+      debug!("Watch thread exiting.");
+      // Signal that we're exiting (which we would also do by just dropping the channel).
+      let _ = thread_liveness_sender.send(());
+    });
+
+    Ok(InvalidationWatcher {
+      watcher,
+      liveness: thread_liveness_receiver,
+    })
+  }
+
+  ///
+  /// Watch the given path non-recursively.
+  ///
+  pub fn watch(&self, path: PathBuf) -> BoxFuture<(), ()> {
+    let watcher = self.watcher.clone();
+    let path2 = path.clone();
+    futures::lazy(move || watcher.lock().watch(path, RecursiveMode::NonRecursive))
+      .then(move |r| match r {
+        Ok(_) => futures::future::ok(()).to_boxed(),
+        Err(e) => {
+          warn!("Failed to watch path {:?}, with error {:?}", path2, e);
+          futures::future::err(()).to_boxed()
+        }
+      })
+      .to_boxed()
+  }
+
+  ///
+  /// Returns true if this InvalidationWatcher is still valid: if it is not valid, it will have
+  /// already logged some sort of error, and will never restart on its own.
+  ///
+  pub fn running(&self) -> bool {
+    match self.liveness.try_recv() {
+      Ok(()) | Err(TryRecvError::Disconnected) => false,
+      Err(TryRecvError::Empty) => true,
+    }
+  }
+
+  pub fn invalidate(graph: &Graph<NodeKey>, paths: &HashSet<PathBuf>, caller: &str) -> usize {
+    let InvalidationResult { cleared, dirtied } = graph.invalidate_from_roots(move |node| {
+      if let Some(fs_subject) = node.fs_subject() {
+        paths.contains(fs_subject)
```

Comment:
I _think_ this will over-invalidate significantly, e.g. it will invalidate the `ScanDir` of the build root for all files. - I would guess that we only want to invalidate if `path == fs_subject` or `Some(path) == fs_subject.parent()`

We should make sure to have a test for this :)

Reply:
I think you meant to comment on the `parent` part of the code: that is an existing pattern copied from elsewhere, because watchman did not allow us to differentiate creates/renames from content changes (so we needed to invalidate the containing directory listing when a file changed: see #4999). When watchman is gone, I think that we might be able to use the richer information that we get from `notify` to improve our precision.

Reply:
I actually just realized that we can use the Create/Remove vs. Modify events in notify to do this now. I need to add back some code to the watchman code path, but that should be fine.

> it will invalidate the ScanDir of the build root for all files

The `paths` set here contains the file that changed, and maybe its parent dir, the fs_subject is just a single path https://github.com/pantsbuild/pants/blob/0afb8a617cd91212222c87796c088ec7d49a2090/src/rust/engine/src/nodes.rs#L1010
So unless the file was a child of the build root it wouldn't invalidate a scandir on the build root.

Reply:
@hrfuller : We should do that, but not in this PR, IMO. Let's focus on parity, and then once the other implementation is removed, we can lean in to make them more complex.

Reply:
On over-invalidation: You're right! I was reading this as `Path::contains` not `HashSet::contains`! Apologies!